crates/unleash-edge-feature-refresh/src/delta_refresh.rs

```diff
@@ -32,129 +35,199 @@ pub type Environment = String;
 
 const DELTA_CACHE_LIMIT: usize = 100;
 
+type SseStream = Pin<
+    Box<
+        dyn futures::Stream<Item = Result<eventsource_client::SSE, eventsource_client::Error>>
+            + Send,
+    >,
+>;
+
+fn reconnect_opts() -> eventsource_client::ReconnectOptions {
+    eventsource_client::ReconnectOptions::reconnect(true)
+        .retry_initial(true)
+        .delay(Duration::from_secs(5))
+        .delay_max(Duration::from_secs(30))
+        .backoff_factor(2)
+        .build()
+}
+
+fn build_sse_stream(
+    streaming_url: &str,
+    token: &EdgeToken,
+    client_meta_information: &ClientMetaInformation,
+    custom_headers: &[(String, String)],
+) -> anyhow::Result<SseStream> {
+    let mut es_client_builder = eventsource_client::ClientBuilder::for_url(streaming_url)
+        .context("Failed to create EventSource client for streaming")?
+        .header("Authorization", &token.token)?
+        .header(UNLEASH_APPNAME_HEADER, &client_meta_information.app_name)?
+        .header(
+            UNLEASH_INSTANCE_ID_HEADER,
+            &client_meta_information.instance_id.to_string(),
+        )?
+        .header(
+            UNLEASH_CONNECTION_ID_HEADER,
+            &client_meta_information.connection_id.to_string(),
+        )?
+        .header(
+            UNLEASH_CLIENT_SPEC_HEADER,
+            unleash_yggdrasil::SUPPORTED_SPEC_VERSION,
+        )?;
+
+    for (key, value) in custom_headers {
+        es_client_builder = es_client_builder.header(key, value)?;
+    }
+
+    let client = es_client_builder.reconnect(reconnect_opts()).build();
+    Ok(client.stream())
+}
+
+async fn handle_sse(
+    sse: eventsource_client::SSE,
+    delta_refresher: &Arc<DeltaRefresher>,
+    token: &EdgeToken,
+) {
+    match sse {
+        eventsource_client::SSE::Event(event)
+            if event.event_type == "unleash-connected" || event.event_type == "unleash-updated" =>
+        {
+            match event.event_type.as_str() {
+                "unleash-connected" => {
+                    debug!("Connected to unleash! Populating my flag cache now.")
+                }
+                "unleash-updated" => debug!("Got an unleash updated event. Updating cache."),
+                _ => unreachable!(),
+            }
```

Comment:
Just to keep the original logs, but I'm also okay with dropping this if you think we don't need it. We can add a single log that is generic for both "connecting" and "updating" if you prefer.

Reply:
Yeah this is totally fine, if you want something slightly more elegant than the match + unreachable, an if statement would probably be my go to here:

``` rust
    if event.event_type == "unleash-connected" {
        debug!("Connected to unleash! Populating flag cache now.");
    } else {
        debug!("Got an unleash updated event. Updating cache.");
    }
```

Reply:
Of course 🤦 in my attempt to write idiomatic Rust I sometimes end up overcomplicating.